### ISSUE TITLE
Pages are matched by page name - cannot export 2 sphinxes to one confluence space

### ISSUE THREAD

**Issue by jakkul** [sphinx-contrib/confluencebuilder#259]
Let's imagine:

- I have a Confluence space, where I want all my Sphinx documents to be published. Let's say that this is the only confluence space that my users look at, the rest is internal.
- I have one Sphinx document per one piece of software (say documentation for system that processes images, documentation for a system that processes books, documentation for system that holds metadata for both).
- Each Sphinx document follows the same architecture with Intro, Disclaimers, Architecture overview, How to install, How to configure, Using, etc, Release Notes.

So now when I deploy these documents, individual pages are identified by the name of the page, so if the first thing I export has a page named "release notes", when I export a second sphinx document with "release notes" in it, it will overwrite the first page on confluence. 

Is there a workaround for this behaviour other than making sure I have the chapters named with individual names? Or maybe the export could build a map with IDs and store it locally with the sphinx docs?

**Comment by jdknight** [comment]
@jakkul, one could say the primary limitation exists in Confluence where (as you have observed), page names must be unique within a space. There exists the [`confluence_publish_prefix`](https://sphinxcontrib-confluencebuilder.readthedocs.io/en/latest/configuration.html#confluence-publish-prefix) and [`confluence_publish_postfix`](https://sphinxcontrib-confluencebuilder.readthedocs.io/en/latest/configuration.html#confluence-publish-postfix) options which can help you work around this issue. A detailed example of the issue (for reference) would be the tutorial's ["confluence spaces and unique page names"](https://sphinxcontrib-confluencebuilder.readthedocs.io/en/latest/tips.html#confluence-spaces-and-unique-page-names) section.

That being said, this does not really help publishers which do not necessarily want to use prefixes/post-fixes in their published set but also do not have control over the content being published (i.e. cannot adjust title names). It would be nice to provided a configuration entry which a publisher can override a target document identifier to a title value (over a detected or missing title entry). Is such a feature desirable?

**Comment by jakkul** [comment]
Thank you for these options - I'll try to use them, these should do the trick!

I do not fully understand the second to last sentence. I suppose you are suggesting adding a feature that will map the confluence page names to some unique identifiers - I might be wrong. This is something more than welcome. But if what you are trying to do here is to provide another way of working around Confluence limitations, then maybe just let people use the postfixes and suffixes and code something more important :)

One thing might be needed - if the script detects a page with the same name already there on the COnfluence space that is not nested where a previous version of the same name should be, maybe there should be a warning or even the export should stop?

**Comment by jakkul** [comment]
by the way, thank you for the effort to make such an export tool!

**Comment by jdknight** [comment]
> I suppose you are suggesting adding a feature that will map the confluence page names to some unique identifiers ... [this] is something more than welcome.

That is indeed my suggestion. My initial thoughts on this is just adding a dictionary entry to the configuration which allows a publisher to override the detected title name if a docname is matched. I believe this would provide publishers a more complete solution (with existing workarounds) to deal with document title assignments.

> [If] the script detects a page with the same name already there on the [Confluence] space that is not nested where a previous version of the same name should be, maybe there should be a warning or even the export should stop?

Hmm... something like this should be possible. I do not think I would be a fan of stopping a publish event outright, but we could try to generate more warnings when a possible "override" is detected. When publishing a document, this extension does not specifically track where pages were located. Page identifiers are tracked and uses in a waterfall manner (i.e. in a crude way: publish base document, track base ID, publish next document based on that ID). We could adjust the publish event to fetch possible ancestors (unfortunately, an extra REST call per document) and compare with the target parent identifier that will be set:

```
def publish_doc(self, docname, output):
    ...
    original_page_ancestors = None
    _, original_page = self.publisher.getPage(title)
    if original_page and 'ancestors' in original_page:
        original_page_ancestors = original_page['ancestors']
    ...
    if original_page_ancestors and parent_id:
        # todo | some check here; trigger warning if a match is not found
```

If something like this is introduced (assuming it actually works and deemed useful), there would also need to be a consideration on how to promote a more performance-based publish process. While this above check helps warn about an issue, for environments which do not experience this (which I assume would be more common), the sanity check is just an unnecessary IO call.

Also -- if we aim to support both the "safe" check and the performance method, what is best by default and how to configure for these.

**Comment by jakkul** [comment]
Hey, prefix does the trick for me, but postfix just does not work on

**Comment by jakkul** [comment]
ok. I've tested the prefix option and it basically solves all of my problems.

Regarding the behaviour of the script at upload and finding pages with the same name but wrong hierarchy - your points are important. And I think that depending on the expectations from the software you might want to tackle them differently. And I guess I'm not in the position to recommend going one way or another. I'm using the software manually, on as needed basis. We have tens of sphinx repos, all of them are being pushed to confluence in the end, but all of this is on as needed basis, manual. No one is using CI to push changes to Confluence on every commit. Yet :)

But there's one thing that you can always do that is really useful regardless of the way the SW will be used. A --dry-run option could be added that will list all the actions that will need to be taken on the wiki to get the site updated. "I've found such and such page and I will update it" "I've found such and such page, but it is nested somewhere funny, so I recommend checking it up before the update" etc.

**Comment by jdknight** [comment]
> but postfix just does not work on

Interesting. I will note and take a look when doing system testing for the upcoming release.

> A --dry-run option could be added

I like it! For the Sphinx-way of doing things, I am not really sure if extensions "should" be accepting configurations via command line arguments (other than possible configuration values passed in via command line). While I think everyone would agree that just `--dry-run` is nice; the "proper" extension way maybe something like `-Dconfluence_dryrun=1`.

**Comment by jdknight** [comment]
My mistake on post-fixes; they are currently only available on `master` at this time and will be available next stable release (I really should update the configuration documentation to reflect which options have been made available in which version).

**Comment by jdknight** [comment]
See also #296 in relation to a dry run capability.

**Comment by jdknight** [comment]
From this issue, it spawned the ideas of adding both a dry run capability (#296) and support explicit title overrides (#297). Both features now exist on the `master` branch (see respective pull requests referenced) and will be available next stable release (v1.3+). For issues or enhancements to these respective issues, it is recommended to open a new issue outlining the use case.

For now, users on stable releases who are experiencing page name conflicts/issues can take advantage of [`confluence_publish_prefix`](https://sphinxcontrib-confluencebuilder.readthedocs.io/en/latest/configuration.html#confluence-publish-prefix) and [`confluence_publish_postfix`](https://sphinxcontrib-confluencebuilder.readthedocs.io/en/latest/configuration.html#confluence-publish-postfix) options; both which are available in the current v1.2 release.

@jakkul, thanks for the discussion. The issue will now be marked as closed.
